samza-core/src/main/java/org/apache/samza/task/TaskFactoryUtil.java

```diff
@@ -147,29 +150,31 @@ private static void validateFactory(Object factory) {
     }
   }
 
-  private static StreamTask createStreamOperatorTask(Config config, ApplicationRunner runner) throws Exception {
-    Class<?> builderClass = Class.forName(config.get(StreamApplication.APP_CLASS_CONFIG));
-    StreamApplication graphBuilder = (StreamApplication) builderClass.newInstance();
-    return new StreamOperatorTask(graphBuilder, runner);
-  }
-
-  private static boolean isStreamOperatorTask(Config config) {
+  /**
+   * Create a {@link StreamApplication} from config.
+   * Throw {@link ConfigException} if the configured class name is not a StreamApplication type, or
+   * the low level task.class is also configured.
+   * @param config Config
+   * @return {@link StreamApplication} instance
+   */
+  public static StreamApplication createStreamApplication(Config config) {
     if (config.get(StreamApplication.APP_CLASS_CONFIG) != null && !config.get(StreamApplication.APP_CLASS_CONFIG).isEmpty()) {
-
       TaskConfig taskConfig = new TaskConfig(config);
       if (taskConfig.getTaskClass() != null && !taskConfig.getTaskClass().isEmpty()) {
         throw new ConfigException("High level StreamApplication API cannot be used together with low-level API using task.class.");
       }
 
+      String appClassName = config.get(StreamApplication.APP_CLASS_CONFIG);
       try {
-        Class<?> builderClass = Class.forName(config.get(StreamApplication.APP_CLASS_CONFIG));
-        return StreamApplication.class.isAssignableFrom(builderClass);
+        Class<?> builderClass = Class.forName(appClassName);
+        return (StreamApplication) builderClass.newInstance();
       } catch (Throwable t) {
         log.error("Failed to validate StreamApplication class from the config. {}={}",
-            StreamApplication.APP_CLASS_CONFIG, config.get(StreamApplication.APP_CLASS_CONFIG));
-        return false;
+            StreamApplication.APP_CLASS_CONFIG, appClassName);
+        throw new ConfigException(String.format("%s is not a StreamApplication.", appClassName));
       }
+    } else {
+      return null;
```

Comment:
It's weird that this both throws an exception and returns null, should probably throw in the else case too.
Also, don't need to both log and throw the exception - we can throw the exception with an informative message here and make sure it gets caught and logged upstream.